File: PdfReader.java
// PdfReader.java
// Версия на Java с использованием Apache PDFBox

import org.apache.pdfbox.pdmodel.PDDocument;
import org.apache.pdfbox.pdmodel.PDDocumentInformation;
import org.apache.pdfbox.text.PDFTextStripper;
import org.apache.pdfbox.Loader;

import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;

public class PdfReader {
    private final String filename;
    private final String password;
    private PDDocument document;

    public PdfReader(String filename, String password) throws IOException {
        this.filename = filename;
        this.password = password;
        File file = new File(filename);
        if (password != null && !password.isEmpty()) {
            document = Loader.loadPDF(file, password);
        } else {
            document = Loader.loadPDF(file);
        }
    }

    public void close() throws IOException {
        if (document != null) document.close();
    }

    public Map<String, String> getMetadata() {
        PDDocumentInformation info = document.getDocumentInformation();
        Map<String, String> meta = new LinkedHashMap<>();
        meta.put("Title", info.getTitle());
        meta.put("Author", info.getAuthor());
        meta.put("Creator", info.getCreator());
        meta.put("Producer", info.getProducer());
        meta.put("CreationDate", info.getCreationDate() != null ? info.getCreationDate().toString() : "");
        meta.put("ModDate", info.getModificationDate() != null ? info.getModificationDate().toString() : "");
        return meta;
    }

    public int getPageCount() {
        return document.getPages().getCount();
    }

    public String extractText(int pageNum) throws IOException {
        PDFTextStripper stripper = new PDFTextStripper();
        if (pageNum > 0) {
            stripper.setStartPage(pageNum);
            stripper.setEndPage(pageNum);
        }
        return stripper.getText(document);
    }

    public Map<Integer, List<String>> search(String query, int pageNum) throws IOException {
        String text = extractText(pageNum);
        String[] lines = text.split("\n");
        Map<Integer, List<String>> results = new LinkedHashMap<>();
        for (int i = 0; i < lines.length; i++) {
            if (lines[i].toLowerCase().contains(query.toLowerCase())) {
                results.computeIfAbsent(i+1, k -> new ArrayList<>()).add(lines[i].trim());
            }
        }
        return results;
    }

    // ANSI-цвета (консольные)
    private static String colorize(String text, String color) {
        return color + text + "\u001B[0m";
    }

    public static void main(String[] args) {
        String filename = null;
        String password = null;
        boolean textFlag = false;
        boolean metadataFlag = false;
        boolean pagesFlag = false;
        String searchQuery = null;
        String outputFile = null;
        int pageNum = 0;

        // Простой парсинг аргументов
        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "-t": textFlag = true; break;
                case "-m": metadataFlag = true; break;
                case "-p": pagesFlag = true; break;
                case "-s": if (i+1 < args.length) searchQuery = args[++i]; break;
                case "-o": if (i+1 < args.length) outputFile = args[++i]; break;
                case "--password": if (i+1 < args.length) password = args[++i]; break;
                case "--page": if (i+1 < args.length) pageNum = Integer.parseInt(args[++i]); break;
                default:
                    if (!args[i].startsWith("-")) filename = args[i];
            }
        }

        if (filename == null) {
            System.err.println("Укажите PDF файл.");
            System.exit(1);
        }

        try (PdfReader reader = new PdfReader(filename, password)) {
            if (metadataFlag) {
                Map<String, String> meta = reader.getMetadata();
                System.out.println(colorize("Метаданные:", "\u001B[1m"));
                for (Map.Entry<String, String> e : meta.entrySet()) {
                    System.out.printf("  %s: %s\n", colorize(e.getKey(), "\u001B[96m"), e.getValue());
                }
                System.out.println();
            }

            if (pagesFlag) {
                System.out.printf("Количество страниц: %s\n", colorize(String.valueOf(reader.getPageCount()), "\u001B[92m"));
                System.out.println();
            }

            if (searchQuery != null) {
                Map<Integer, List<String>> results = reader.search(searchQuery, pageNum);
                if (!results.isEmpty()) {
                    System.out.printf("%s\n", colorize("Найдено совпадений: " + results.size(), "\u001B[92m"));
                    for (Map.Entry<Integer, List<String>> e : results.entrySet()) {
                        for (String line : e.getValue()) {
                            String highlighted = line.replaceAll("(?i)" + searchQuery, colorize("$0", "\u001B[93m"));
                            System.out.printf("  Строка %d: %s\n", e.getKey(), highlighted);
                        }
                    }
                } else {
                    System.out.println(colorize("Совпадений не найдено.", "\u001B[91m"));
                }
                System.out.println();
            }

            if (textFlag || (!metadataFlag && !pagesFlag && searchQuery == null)) {
                String text = reader.extractText(pageNum);
                if (outputFile != null) {
                    Files.write(Paths.get(outputFile), text.getBytes());
                    System.out.println(colorize("Текст сохранён в " + outputFile, "\u001B[92m"));
                } else {
                    System.out.println(colorize("Извлечённый текст:", "\u001B[1m"));
                    System.out.println(text);
                }
            }
        } catch (Exception e) {
            System.err.println(colorize("Ошибка: " + e.getMessage(), "\u001B[91m"));
            System.exit(1);
        }
    }
}

File: PdfReader.cs
// PdfReader.cs
// Версия на C# с использованием PdfPig, System.CommandLine (или простой парсинг)

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using UglyToad.PdfPig;
using UglyToad.PdfPig.Content;

namespace PdfReader
{
    class Program
    {
        // ANSI-цвета
        const string Reset = "\u001b[0m";
        const string Cyan = "\u001b[96m";
        const string Green = "\u001b[92m";
        const string Yellow = "\u001b[93m";
        const string Red = "\u001b[91m";
        const string Blue = "\u001b[94m";
        const string Bold = "\u001b[1m";

        static string Colorize(string text, string color) => color + text + Reset;

        static void Main(string[] args)
        {
            string filename = null;
            string password = null;
            bool textFlag = false;
            bool metadataFlag = false;
            bool pagesFlag = false;
            string searchQuery = null;
            string outputFile = null;
            int pageNum = 0;

            for (int i = 0; i < args.Length; i++)
            {
                switch (args[i])
                {
                    case "-t": textFlag = true; break;
                    case "-m": metadataFlag = true; break;
                    case "-p": pagesFlag = true; break;
                    case "-s": if (i + 1 < args.Length) searchQuery = args[++i]; break;
                    case "-o": if (i + 1 < args.Length) outputFile = args[++i]; break;
                    case "--password": if (i + 1 < args.Length) password = args[++i]; break;
                    case "--page": if (i + 1 < args.Length) pageNum = int.Parse(args[++i]); break;
                    default:
                        if (!args[i].StartsWith("-")) filename = args[i];
                        break;
                }
            }

            if (filename == null)
            {
                Console.WriteLine("Укажите PDF файл.");
                return;
            }

            try
            {
                using (var pdf = PdfDocument.Open(filename, new ParsingOptions { Password = password }))
                {
                    if (metadataFlag)
                    {
                        Console.WriteLine(Colorize("Метаданные:", Bold));
                        var info = pdf.Information;
                        var meta = new Dictionary<string, string>
                        {
                            ["Title"] = info.Title,
                            ["Author"] = info.Author,
                            ["Creator"] = info.Creator,
                            ["Producer"] = info.Producer,
                            ["CreationDate"] = info.CreationDate?.ToString() ?? "",
                            ["ModDate"] = info.ModifiedDate?.ToString() ?? ""
                        };
                        foreach (var kv in meta)
                            Console.WriteLine($"  {Colorize(kv.Key, Cyan)}: {kv.Value}");
                        Console.WriteLine();
                    }

                    if (pagesFlag)
                    {
                        Console.WriteLine($"Количество страниц: {Colorize(pdf.NumberOfPages.ToString(), Green)}");
                        Console.WriteLine();
                    }

                    if (searchQuery != null)
                    {
                        var results = new Dictionary<int, List<string>>();
                        int startPage = pageNum > 0 ? pageNum : 1;
                        int endPage = pageNum > 0 ? pageNum : pdf.NumberOfPages;
                        for (int p = startPage; p <= endPage; p++)
                        {
                            var page = pdf.GetPage(p);
                            string text = page.Text;
                            var lines = text.Split('\n');
                            for (int i = 0; i < lines.Length; i++)
                            {
                                if (lines[i].ToLower().Contains(searchQuery.ToLower()))
                                {
                                    if (!results.ContainsKey(p)) results[p] = new List<string>();
                                    results[p].Add(lines[i].Trim());
                                }
                            }
                        }
                        if (results.Count > 0)
                        {
                            Console.WriteLine(Colorize($"Найдено совпадений: {results.Values.Sum(l => l.Count)}", Green));
                            foreach (var kv in results)
                            {
                                foreach (var line in kv.Value)
                                {
                                    var highlighted = Regex.Replace(line, searchQuery, m => Colorize(m.Value, Yellow), RegexOptions.IgnoreCase);
                                    Console.WriteLine($"  Страница {kv.Key}, строка: {highlighted}");
                                }
                            }
                        }
                        else
                        {
                            Console.WriteLine(Colorize("Совпадений не найдено.", Red));
                        }
                        Console.WriteLine();
                    }

                    if (textFlag || (!metadataFlag && !pagesFlag && searchQuery == null))
                    {
                        var sb = new StringBuilder();
                        int start = pageNum > 0 ? pageNum : 1;
                        int end = pageNum > 0 ? pageNum : pdf.NumberOfPages;
                        for (int p = start; p <= end; p++)
                        {
                            var page = pdf.GetPage(p);
                            sb.Append(page.Text);
                        }
                        string fullText = sb.ToString();
                        if (outputFile != null)
                        {
                            File.WriteAllText(outputFile, fullText);
                            Console.WriteLine(Colorize($"Текст сохранён в {outputFile}", Green));
                        }
                        else
                        {
                            Console.WriteLine(Colorize("Извлечённый текст:", Bold));
                            Console.WriteLine(fullText);
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine(Colorize($"Ошибка: {ex.Message}", Red));
            }
        }
    }
}
